Add well-known-binary decoder for geometry and geography

platypii opened the pull request:
Adds hyparquet support for decoding parquet GEOMETRY and GEOGRAPHY types into geojson compatible Geometry objects. [RFC7946](https://datatracker.ietf.org/doc/html/rfc7946)

Adds `geometryFromBytes` and `geographyFromBytes` to the parsers object so the parsing can be overriden by users.

This includes improvements to the WKB implementation in [geoparquet](https://github.com/hyparam/geoparquet). This implementation also supports:
 - GeometryCollections
 - Multi-dimensional points via [Extended WKB](https://libgeos.org/specifications/wkb/#extended-wkb)

Includes `geospatial.parquet` test from `parquet-testing` repo.

severo commented on test/wkb.test.js:
```diff
@@ -0,0 +1,283 @@
+    expect(() => decodeWKB(makeReader(buffer))).toThrowError('Unsupported geometry type: 99')
+  })
+
+  it('decodes EWKB Point with SRID and Z/M flags', () => {
```
Should we expose the SRID instead of skipping it?

severo commented on src/wkb.js:
```diff
@@ -0,0 +1,206 @@
+    reader.offset += 4
+    const coords = new Array(numRings)
+    for (let r = 0; r < numRings; r++) {
+      const numPoints = view.getUint32(reader.offset, isLittleEndian)
```
the standard expects the number of points to be at least 4, and the first and last points must be the same. Should we enforce it? Or is it ok to produce invalid geojson if the data is not well formed?

platypii replied:
If the input file is invalid, then it is [Undefined Behavior](https://en.wikipedia.org/wiki/Undefined_behavior).
My philosophy is that we should not validate or enforce correctness unnecessarily.

platypii commented on test/files/geospatial.json:
```diff
@@ -0,0 +1,1142 @@
+      "coordinates": [
+        30,
+        10,
+        300
```
The problem is that if hyparquet strips coordinates, then there will be no way for users to access any higher-dimensional coordinates from a parquet file. I think we should output all the coordinates from hyparquet. The spec says "should not" not "must not" :smiley:

platypii replied:
I could be convinced to truncate coordinates on a proper "toGeoJson" function. But I'm not sure it's necessary?

severo replied:
As I said, I think we should not strip them. I just wanted to ask, as it is unclear.

platypii commented:
@severo I addressed most of your comments. I'm not convinced on the fancy types, I think that could cause more type issues than the benefit it gives of asserting that the list is at least length 2. But let me know your thoughts.

severo commented on src/wkb.js:
```diff
@@ -0,0 +1,127 @@
+
+  if (flags.type === 2) { // LineString
+    return { type: 'LineString', coordinates: readLine(reader, flags) }
+  } else if (flags.type === 3) { // Polygon
```
nit: the coding style changes here.

if - return
if - return

but now:

else if - return
else if - return

severo commented on src/wkb.js:
```diff
@@ -0,0 +1,127 @@
+  if (flags) dim++
+  if (flags === 3) dim++
+
+  return { littleEndian, type, dim, count }
```
just to be sure: once decoded, there is no way to tell if it was XYZ or XYM, right? If I understand well, GeoJSON does not give a way to know?

severo replied:
Yes, I think the best is to return geojson. Not sure if Z / M are used too much, and it's not our faukt if it's not supported well in geojson

severo commented on test/wkb.test.js:
```diff
@@ -0,0 +1,211 @@
+    expect(() => wkbToGeojson(makeReader(buffer))).toThrowError('Unsupported geometry type: 99')
+  })
+
+  it('decodes EWKB Point with Z/M flags', () => {
```
should it be ISO WKB instead of EWKB?
